include/boost/json/impl/serialize.ipp

```diff
@@ -59,62 +61,173 @@ serialize_impl(
 
 std::string
 serialize(
-    value const& jv)
+    value const& jv,
+    const serialize_options & opt)
 {
     std::string s;
     serializer sr;
     sr.reset(&jv);
-    serialize_impl(s, sr);
+    serialize_impl(s, sr, opt);
     return s;
 }
 
 std::string
 serialize(
-    array const& arr)
+    array const& arr,
+    const serialize_options & opt)
 {
     std::string s;
     serializer sr;
     sr.reset(&arr);
-    serialize_impl(s, sr);
+    serialize_impl(s, sr, opt);
     return s;
 }
 
 std::string
 serialize(
-    object const& obj)
+    object const& obj,
+    const serialize_options & opt)
 {
     std::string s;
     serializer sr;
     sr.reset(&obj);
-    serialize_impl(s, sr);
+    serialize_impl(s, sr, opt);
     return s;
 }
 
 std::string
 serialize(
-    string const& str)
+    string const& str,
+    const serialize_options & opt)
 {
     std::string s;
     serializer sr;
     sr.reset(&str);
-    serialize_impl(s, sr);
+    serialize_impl(s, sr, opt);
     return s;
 }
 
 // this is here for key_value_pair::key()
 std::string
 serialize(
-    string_view sv)
+    string_view sv,
+    const serialize_options & opt)
 {
     std::string s;
     serializer sr;
     sr.reset(sv);
-    serialize_impl(s, sr);
+    serialize_impl(s, sr, opt);
     return s;
 }
 
 //----------------------------------------------------------
 
+
+static
+void
+serialize_impl(
+    std::string& s,
+    serializer& sr,
+    error_code & ec)
+{
+  // serialize to a small buffer to avoid
+  // the first few allocations in std::string
+  char buf[BOOST_JSON_STACK_BUFFER_SIZE];
```

Comment:
Why are we duplicating this function?

Reply:
I thought it would turn out different, i.e. need some early breaks for errors. Was wrong, could be redone as a template.

Reply:
A template would still result in duplication in the emitted code, we need to avoid that since Boost.JSON needs to be as lean as possible.